caching_file_reader.go

```diff
@@ -0,0 +1,68 @@
+package kubeauth
+
+import (
+	"io/ioutil"
+	"sync"
+	"time"
+)
+
+// cachingFileReader reads a file and keeps an in-memory copy of it, until the
+// copy is considered stale. Next ReadFile() after expiry will re-read the file from disk.
+type cachingFileReader struct {
+	// path is the file path to the cached file.
+	path string
+
+	// ttl is the time-to-live duration when cached file is considered stale
+	ttl time.Duration
+
+	// cache is the buffer holding the in-memory copy of the file.
+	cache *cachedFile
+
+	l sync.RWMutex
+
+	// currentTime is a function that returns the current local time.
+	// Normally set to time.Now but it can be overwritten by test cases to manipulate time.
+	currentTime func() time.Time
+}
+
+type cachedFile struct {
+	// buf is the buffer holding the in-memory copy of the file.
+	buf string
+
+	// expiry is the time when the cached copy is considered stale and must be re-read.
+	expiry time.Time
+}
+
+func newCachingFileReader(path string, ttl time.Duration, currentTime func() time.Time) *cachingFileReader {
+	return &cachingFileReader{
+		path:        path,
+		ttl:         ttl,
+		currentTime: currentTime,
+	}
+}
+
+func (r *cachingFileReader) ReadFile() (string, error) {
+	// Fast path requiring read lock only: file is already in memory and not stale.
+	r.l.RLock()
+	now := r.currentTime()
+	cache := r.cache
```

Comment:
Since `r.cache` is a pointer value, I think we'd have to perform the `cache` nil-check and `cache.expiry` check while the lock is held to avoid possible mutation on the underlying value. I don't think assigning it this way works unless you store the dereferenced pointer value.

Reply:
True, thanks for spotting! I changed it from pointer to value instead.